File: src/minecraft/net/machinemuse/powersuits/network/packets/MusePacketModeChangeRequest.java
/**
 *
 */
package net.machinemuse.powersuits.network.packets;

import cpw.mods.fml.common.network.Player;
import net.machinemuse.powersuits.network.MusePacket;
import net.machinemuse.utils.MuseItemUtils;
import net.minecraft.client.entity.EntityClientPlayerMP;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;

import java.io.DataInputStream;
import java.util.List;

/**
 * @author Claire
 */
public class MusePacketModeChangeRequest extends MusePacket {
    protected String mode;
    protected int slot;

    /**
     */
    public MusePacketModeChangeRequest(Player player, String mode, int slot) {
        super(player);
        writeInt(slot);
        writeString(mode);
    }

    /**
     * @param player
     * @param data
     */
    public MusePacketModeChangeRequest(DataInputStream data, Player player) {
        super(data, player);
        slot = readInt();
        mode = readString(64);
    }

    @Override
    public void handleClient(EntityClientPlayerMP player) {

    }

    @Override
    public void handleServer(EntityPlayerMP player) {
        ItemStack stack = null;
        if (slot > -1 && slot < 9) {
            // player.inventory.currentItem = slot;
            stack = player.inventory.mainInventory[slot];
        }

        if (stack == null) {
            return;
        }
        List<String> modes = MuseItemUtils.getModes(stack, player);
        if (modes.contains(mode)) {
            NBTTagCompound itemTag = MuseItemUtils.getMuseItemTag(stack);
            itemTag.setString("Mode", mode);
        }
    }

}

File: src/minecraft/net/machinemuse/powersuits/network/packets/MusePacketModeChangeRequest.scala
/**
 *
 */
package net.machinemuse.powersuits.network.packets

import cpw.mods.fml.common.network.Player
import net.machinemuse.utils.MuseItemUtils
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.item.ItemStack
import java.io.DataInputStream
import java.util.List
import scala.Predef._
import net.machinemuse.numina.network.{MusePackager, MusePacket}
import net.machinemuse.numina.item.ModeChangingItem
import net.machinemuse.numina.scala.OptionCast
import net.machinemuse.powersuits.item.ModeChangingModularItem

/**
 * Author: MachineMuse (Claire Semple)
 * Created: 12:28 PM, 5/6/13
 */
object MusePacketModeChangeRequest extends MusePackager {
  def read(d: DataInputStream, p: Player) = {
    val slot = readInt(d)
    val mode = readString(d)
    new MusePacketModeChangeRequest(p, mode, slot)
  }
}

class MusePacketModeChangeRequest(player: Player, mode: String, slot: Int) extends MusePacket(player) {
  val packager = MusePacketModeChangeRequest

  def write {
    writeInt(slot)
    writeString(mode)
  }

  override def handleServer(player: EntityPlayerMP) {
    var stack: ItemStack = null
    if (slot > -1 && slot < 9) {
      for {
        stack <- Option(player.inventory.mainInventory(slot))
        item <- OptionCast[ModeChangingModularItem](stack.getItem)
      } {
        val modes = item.getModes(stack)
        if (modes.contains(mode)) {
          OptionCast[ModeChangingItem](stack.getItem).map(i => i.setActiveMode(stack, mode))
        }
      }
    }
  }

}
